src/components/learner-credit-management/data/utils.js

```diff
@@ -551,3 +555,97 @@ export const isLmsBudget = (
   activeIntegrationsLength,
   isUniversalGroup,
 ) => activeIntegrationsLength > 0 && isUniversalGroup;
+
+/**
+ * Determines if the course has already started. Mostly used around text formatting for tense
+ *
+ * @param start
+ * @returns {boolean}
+ */
+export const hasCourseStarted = (start) => dayjs(start).isBefore(dayjs());
+
+/**
+ * Returns assignable course runs within the threshold of within the subsidies expiration date
+ * offset by the DAYS_UNTIL_ASSIGNMENT_ALLOCATION_EXPIRATION constant. It sorts it from the soonest expiring
+ * enroll-by date and the enroll-by date and upgrade deadline has been normalized to ISO format.
+ *
+ * @param courseRuns
+ * @param subsidyExpirationDatetime
+ * @param isLateRedemptionAllowed
+ * @returns {*}
+ */
+export const getAssignableCourseRuns = ({ courseRuns, subsidyExpirationDatetime, isLateRedemptionAllowed }) => {
+  const clonedCourseRuns = courseRuns.map(courseRun => ({
+    ...courseRun,
+    enrollBy: dayjs.unix(courseRun.enrollBy).toISOString(),
+    upgradeDeadline: dayjs.unix(courseRun.upgradeDeadline).toISOString(),
+  }));
+  const assignableCourseRunsFilter = ({ enrollBy, isActive }) => {
+    const enrollByDateThreshold = dayjs(enrollBy).isBefore(
+      Math.max(
+        dayjs(subsidyExpirationDatetime).subtract(MAX_ALLOWABLE_REFUND_THRESHOLD_DAYS, 'days').toDate(),
+        dayjs().add(DAYS_UNTIL_ASSIGNMENT_ALLOCATION_EXPIRATION, 'days').toDate(),
+      ),
+    );
+    if (isLateRedemptionAllowed) {
+      return enrollByDateThreshold;
+    }
```

Comment:
Unfortunately, I think if we just completely ignore isActive, it would potentially expose runs that are unenrollable due to not being marketable, having an "Archived" availability, or not having a "published" status. Unfortunately any of these could be the case regardless of proximity to the enrollment deadline. We might have to bypass isActive and pull in those 3 checks into the frontend, or extract out a new field in the algolia indexing logic, like `is_marketable_published_unarchived" alongside the existing `is_active` property: https://github.com/openedx/enterprise-catalog/blob/555e018ef7a1e5c701642e729bbd2fd5c8a086bc/enterprise_catalog/apps/catalog/algolia_utils.py#L1111

Reply:
Furthermore, we might just have to completely drop returning `enrollByDateThreshold` in the late redemption case. The entire point of late redemption is to be able to assign/redeem a course *after* the enrollment deadline, let alone after 90 days before the enrollment deadline 😛

Reply:
Given the upcoming PR, we are looking at an explicit `isLateEnrollmentEligibile` field which mimics the derived value from the learner portal [here](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/f1887e42b7fc39623b56184c17258ef9a3f578fa/src/components/app/data/utils.js#L523-L531).

Reply:
[inform/context] Here is the upcoming PR @brobro10000 mentioned: https://github.com/openedx/enterprise-catalog/pull/945 😄 

> Unfortunately, I think if we just completely ignore isActive, it would potentially expose runs that are unenrollable due to not being marketable, having an "Archived" availability, or not having a "published" status.

Given the Learner Portal filtering we in place today for the late enrollment use case ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/f1887e42b7fc39623b56184c17258ef9a3f578fa/src/components/app/data/utils.js#L524-L531)), I believe the only check we're missing here is around the `availability="Archived"` case; great catch.

In the above linked enterprise-catalog PR, we're now going to expose a `is_late_enrollment_eligible` field to take into account the same checks the Learner Portal filter is doing as described above.

> For an MVP and testing purposes, this might be enough, but to be completely accurate we'd still need to loop back and add a add checks to mirror the whole "late enrollment buffer" logic from the learner portal in order to prevent showing really old runs.

The Algolia indexing of the `course_runs` list on top-level course objects includes the "late enrollment buffer" logic to not index really old runs (see above linked enterprise-catalog PR) that are well beyond the late enrollment buffer window. That said, there is the off-chance we could have a stale Algolia index whereby already-indexed courses that were eligible for late enrollment at the time of last index are _no longer_ eligible.

@brobro10000 is going to push another commit to this PR to ensure a similar check as done during Algolia indexing around "Is the `enroll_by` within the late enrollment buffer window?" to account for this edge case of stale Algolia index.